### PR TITLE
Refactor flow tests to use page objects

### LINKED ISSUE #33: Refactor flow tests to use page objects
Reference material:
 - https://github.com/appfolio/ae_page_objects
 - https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects

The starting skeleton is on the `flowTestsWithPageObjects` branch.

### PR DESCRIPTION
**Pull request by lsq921224**
Closes #33

### CODE REVIEW

File: app/views/images/_image.html.erb [hunk 1]
```diff
@@ -1,6 +1,6 @@
-<div class="col-md-4">
+<div class="image-card-container col-md-4">
```
Comment (rmacklin) [inline app/views/images/_image.html.erb]: If this class is not used for styling, you should prefix it with `js-`

Comment (rmacklin) [inline app/views/images/_image.html.erb]: The same goes for all the classes you added here.

File: app/views/images/index.html.erb [hunk 2]
```diff
@@ -1,3 +1,3 @@
-<div class="row">
+<div id="images-list" class="row">
```
Comment (rmacklin) [inline app/views/images/index.html.erb]: `id`s do not need a `js-` prefix (because ids should never be used for styling), but our convention it to use underscores (not hyphens) as the word separator for ids.

File: test/flows/images_crud_test.rb [hunk 3]
```diff
@@ -7,34 +7,35 @@ class ImagesCrudTest < FlowTestCase
     new_image_page = images_index_page.add_new_image!
 
     tags = %w(foo bar)
-    new_image_page = new_image_page.create_image!(url: 'invalid', tags: tags.join(', ')).as_a(PageObjects::Images::NewPage)
-    assert_equal 'must be a valid URL', new_image_page.url.error_message
+    new_image_page = new_image_page.create_image!(link: 'invalid',
+                                                  tags: tags.join(', ')).as_a(PageObjects::Images::NewPage)
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: If you're going to shorten this line, here's the recommended way:
```ruby
new_image_page = new_image_page.create_image!(
  link: 'invalid',
  tags: tags.join(', ')
).as_a(PageObjects::Images::NewPage)
```

Your current way is still quite long and has unnecessary whitespace

File: test/page_objects/images/show_page.rb [hunk 4]
```diff
@@ -3,28 +3,31 @@ module Images
     class ShowPage < PageObjects::Document
       path :image
 
-      def has_url?(url)
-        #TODO
+      collection :tag_elements, locator: '.card-block', item_locator: '.image-tag'
```
Comment (rmacklin) [inline test/page_objects/images/show_page.rb]: Are you using this collection?

File: test/page_objects/images/show_page.rb [hunk 5]
```diff
@@ -3,28 +3,31 @@ module Images
...
-        #TODO
+        node.click_on('delete')
         yield node.driver.browser.switch_to.alert
       end
 
       def delete_and_confirm!
-        #TODO
+        node.click_on('delete')
         alert = node.driver.browser.switch_to.alert
         alert.accept
```
Comment (rmacklin) [inline test/page_objects/images/show_page.rb]: You can refactor this to `delete(&:accept)`

File: app/views/tags/_tag_with_count.html.erb [hunk 6]
```diff
@@ -1,4 +1,4 @@
 <div class="image-tag">
-  <%= link_to tag.name, "#{images_path(tag: tag.name)}" %>
+  <%= link_to tag.name, "#{images_path(tag: tag.name)}", class: 'image-tag__link' %>
```
Comment (rmacklin) [inline app/views/tags/_tag_with_count.html.erb]: As before, if this is not used for styling it should have a `js-` prefix, but additionally it should _not_ follow BEM syntax

File: test/flows/tags_flow_test.rb [hunk 7]
```diff
@@ -2,19 +2,21 @@
...
+                   { link: cute_puppy_url, tag_list: 'puppy, cute' },
+                   { link: cute_puppy_url, tag_list: 'puppy' },
+                   { link: ugly_cat_url, tag_list: 'cat' }])
 
-    assert has_current_path?(images_path, only_path: true), "Does not have #{images_path}"
-    assert page.has_css?('img', count: 1), 'Page should have one image'
-    assert_equal 'http://www.w3schools.com/css/3.jpg', find('img')[:src],
-                 'Page is not showing the correct image'
+    tags_index_page = PageObjects::Tags::IndexPage.visit
+    assert tags_index_page.showing_tags_in_sequence?(tags: %w(puppy cute cat), counts: %w(3 2 1))
```
Comment (rmacklin) [inline test/flows/tags_flow_test.rb]: The interface I'd expect this page object to provide is:
```ruby
assert_equal [['puppy', 3], ['cute', 2], ['cat', 1]], tags_index_page.tags_with_count
```

File: test/flows/tags_flow_test.rb [hunk 8]
```diff
@@ -2,19 +2,21 @@
...
+
+    cat_tag = tags_index_page.tag_list.find do |tag|
+      tag.name == 'cat'
+    end
+
+    images_index_page = cat_tag.view!
+    assert images_index_page.showing_image?(link: ugly_cat_url, tags: %w(cat))
```
Comment (rmacklin) [inline test/flows/tags_flow_test.rb]: You should also assert that the excluded images are _not_ showing.

File: test/flows/signup_login_logout_test.rb [hunk 9]
```diff
@@ -6,81 +6,84 @@ def setup
...
-    find("a[href='#{session_path(User.last)}']").click
-    assert has_current_path?(root_path), "Does not have #{root_path}"
-    assert find('.alert.alert-success').has_text?('Log out successfully.'), 'There should be a success message.'
+    images_index_page = PageObjects::Images::IndexPage.visit
+    new_user_page = images_index_page.sign_up!
+
+    new_user_page = new_user_page.create_user!(name: '',
+                                               email: '',
+                                               password: '',
+                                               password_confirmation: '').as_a(PageObjects::Users::NewPage)
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: Same as before:
```ruby
new_user_page = new_user_page.create_user!(
  name: '',
  email: '',
  password: '',
  password_confirmation: ''
).as_a(PageObjects::Users::NewPage)
```

File: test/flows/signup_login_logout_test.rb [hunk 10]
```diff
@@ -6,81 +6,84 @@ def setup
...
+
+    images_index_page = new_user_page.create_user!(name: name,
+                                                   email: email,
+                                                   password: password,
+                                                   password_confirmation: password)
+    assert_equal 'Account created successfully.', images_index_page.flash_message(:success)
+
+    assert images_index_page.logout_link.present?
+    refute images_index_page.signup_link.present?
+    refute images_index_page.login_link.present?
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: You should use the predicate assertions to make the failure messages better:
```ruby
assert_predicate images_index_page.logout_link, :present?
refute_predicate images_index_page.signup_link, :present?
refute_predicate images_index_page.login_link, :present?
```
which is equivalent to:
```ruby
assert_predicate images_index_page.logout_link, :present?
assert_predicate images_index_page.signup_link, :absent?
assert_predicate images_index_page.login_link, :absent?
```
but `present?` doesn't check for visibility, so this should really be:
```ruby
assert_predicate images_index_page.logout_link, :visible?
assert_predicate images_index_page.signup_link, :absent?
assert_predicate images_index_page.login_link, :absent?
```

File: test/flows/signup_login_logout_test.rb [hunk 11]
```diff
@@ -6,81 +6,84 @@ def setup
...
+
+    images_index_page = images_index_page.log_out!
+    assert_equal 'Log out successfully.', images_index_page.flash_message(:success)
+
+    refute images_index_page.logout_link.present?
+    assert images_index_page.signup_link.present?
+    assert images_index_page.login_link.present?
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: Same as above

File: test/flows/signup_login_logout_test.rb [hunk 12]
```diff
@@ -6,81 +6,84 @@ def setup
...
-
-    find('input[type="submit"]').click
-    assert has_current_path?(images_path), "Does not have #{images_path}"
-    assert find('.alert.alert-success').has_text?('Log in successfully.'), 'There should be a success message.'
-
-    find("a[href='#{session_path(User.last)}']").click
-    assert has_current_path?(root_path), "Does not have #{root_path}"
-    assert find('.alert.alert-success').has_text?('Log out successfully.'), 'There should be a success message.'
+    images_index_page = PageObjects::Images::IndexPage.visit
+    new_user_page = images_index_page.sign_up!
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: Flow tests should use the terminology a user would think in. For example, a user doesn't think "new user page"; they think "sign up page", so our tests and page objects should use names aligned with that.

File: test/flows/signup_login_logout_test.rb [hunk 13]
```diff
@@ -6,81 +6,84 @@ def setup
...
+    refute images_index_page.login_link.present?
+
+    images_index_page = images_index_page.log_out!
+    assert_equal 'Log out successfully.', images_index_page.flash_message(:success)
+
+    refute images_index_page.logout_link.present?
+    assert images_index_page.signup_link.present?
+    assert images_index_page.login_link.present?
+
+    new_session_page = images_index_page.log_in!
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: As above, this should be `login_page`, not `new_session_page`, to match with how users would think about interacting with our application in this flow.

File: test/flows/signup_login_logout_test.rb [hunk 14]
```diff
@@ -6,81 +6,84 @@ def setup
...
-    assert has_current_path?(images_path), "Does not have #{images_path}"
-    assert find('.alert.alert-success').has_text?('Log in successfully.'), 'There should be a success message.'
-
-    find("a[href='#{session_path(User.last)}']").click
-    assert has_current_path?(root_path), "Does not have #{root_path}"
-    assert find('.alert.alert-success').has_text?('Log out successfully.'), 'There should be a success message.'
+    images_index_page = PageObjects::Images::IndexPage.visit
+    new_user_page = images_index_page.sign_up!
+
+    new_user_page = new_user_page.create_user!(name: '',
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: Similarly this should be `sign_up!`, not `create_user!`, which matches the text of the button that's being clicked on.

File: test/flows/signup_login_logout_test.rb [hunk 15]
```diff
@@ -6,81 +6,84 @@ def setup
...
+
+    images_index_page = images_index_page.log_out!
+    assert_equal 'Log out successfully.', images_index_page.flash_message(:success)
+
+    refute images_index_page.logout_link.present?
+    assert images_index_page.signup_link.present?
+    assert images_index_page.login_link.present?
+
+    new_session_page = images_index_page.log_in!
+    new_session_page = new_session_page.create_session!(email: '',
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: Similarly, this should be `log_in!`, not `create_session!`, which matches the text of the button that's being clicked on.

File: test/page_objects/document.rb [hunk 16]
```diff
@@ -1,7 +1,26 @@
...
+    element :signup_link, locator: '.signup-link'
+    element :login_link, locator: '.login-link'
+    element :logout_link, locator: '.logout-link'
+
     def flash_message(message_type)
       node.find(".alert.alert-#{message_type}").text
     end
+
+    def sign_up!
+      node.click_link('Sign up')
```
Comment (rmacklin) [inline test/page_objects/document.rb]: Why aren't you using the `sign_up_link` element here? Same for the other methods.

File: test/page_objects/sessions/new_page.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
+module PageObjects
+  module Sessions
+    class NewPage < PageObjects::Document
```
Comment (rmacklin) [inline test/page_objects/sessions/new_page.rb]: Following the previous discussion, this should be `PageObjects::LoginPage`, rather than `PageObjects::Sessions::NewPage`, to match how users think about the application, not how it's implemented.

File: test/page_objects/users/new_page.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
+module PageObjects
+  module Users
+    class NewPage < PageObjects::Document
```
Comment (rmacklin) [inline test/page_objects/users/new_page.rb]: Following the previous discussion, this should be `PageObjects::SignupPage`, rather than `PageObjects::Users::NewPage`, to match how users think about the application, not how it's implemented.

File: test/flows/signup_login_logout_test.rb [hunk 19]
```diff
@@ -6,81 +6,84 @@ def setup
...
+                                                        password: '',
+                                                        remember_me: 'off').as_a(PageObjects::Sessions::NewPage)
```
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: No need to pass a `remember_me` option if it's not being set.

File: test/page_objects/sessions/new_page.rb [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+      form_for :session do
+        element :email
+        element :password
+        element :remember_me
+      end
+
+      def create_session!(email: nil, password: nil, remember_me: nil)
+        self.email.set email if email.present?
+        self.password.set password if password.present?
+        self.remember_me.set true if remember_me == 'on'
```
Comment (rmacklin) [inline test/page_objects/sessions/new_page.rb]: Let's change the parameter to `remember_me: false` change the interface to just passing `true` when it should be checked, which would change this line to:
```ruby
self.remember_me.set true if remember_me
```

File: test/page_objects/users/new_page.rb [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+        self.email.set email if email.present?
+        self.password.set password if password.present?
+        self.password_confirmation.set password_confirmation if password_confirmation.present?
+
+        click_sign_up!
+      end
+
+      private
+
+      def click_sign_up!
```
Comment (rmacklin) [inline test/page_objects/users/new_page.rb]: What is the purpose of this method?

File: test/flows/signup_login_logout_test.rb (hunk 19 above)
Comment (lsq921224) [inline test/flows/signup_login_logout_test.rb]: In one of my previous PR, I got a comment about passing unused parameters  even if they are not being used, so that the test better reflects what the app is doing. https://github.com/AppFolioOnboarding/image-sharer-lsq921224/pull/30#discussion_r73762460
Even though that comment was on a controller test, I do think it makes sense in flow tests as well, so I added` remember_me` here.

File: test/page_objects/users/new_page.rb (hunk 21 above)
Comment (lsq921224) [inline test/page_objects/users/new_page.rb]: `rubocop` was complaining that `Assignment Branch Condition Size` is too high for `def create_user!(name: nil, email: nil, password: nil, password_confirmation: nil)`, I couldn't find a better way to refactor that method, so I abstracted these two lines of code

File: test/flows/signup_login_logout_test.rb (hunk 19 above)
Comment (rmacklin) [inline test/flows/signup_login_logout_test.rb]: > Even though that comment was on a controller test, I do think it makes sense in flow tests as well

Respectfully, I completely disagree. In a controller test it makes _perfect sense_ to make your requests match what requests the actual application will make. But in a flow tests, you should match what a user would actually do (which in this case is: completely ignore the remember_me checkbox). The request the controller receives will still be the same.

File: test/page_objects/users/new_page.rb (hunk 21 above)
Comment (rmacklin) [inline test/page_objects/users/new_page.rb]: I don't see how extracting these two non-conditional lines would fix a "too many conditions" error.

Comment (lsq921224) [inline test/page_objects/users/new_page.rb]: well, I think it helps reduce `branch`, http://c2.com/cgi/wiki?AbcMetric, as a result it fixes this error

File: test/flows/signup_login_logout_test.rb (hunk 19 above)
Comment (lsq921224) [inline test/flows/signup_login_logout_test.rb]: I see

File: test/page_objects/users/new_page.rb (hunk 21 above)
Comment (rmacklin) [inline test/page_objects/users/new_page.rb]: I have to say I just don't agree that this method's complexity is any better with the private method. I think it's worse this way. Rubocop is great, but I'm not a fan of making code less direct/readable just to silence its complaint. If you wish to keep this rule configured as is, I think a better abstraction would be to extract a method that DRYs the repeated pattern:
```ruby
self.<thing>.set <thing> if <thing>.present?
```

However, I honestly think the method is just fine without extracting anything.

Comment (bboe) [inline test/page_objects/users/new_page.rb]: @lsq921224 please add this file to your repo: https://github.com/AppFolioOnboarding/base/blob/master/test/.rubocop.yml

Some of the metrics don't make sense to have in the test code.

**Comment by lsq921224** [comment]
@rmacklin I think I addressed your comments and I updated the PR.

File: test/page_objects/tags/index_page.rb [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+
+      collection :tag_list, locator: '#tags-list', item_locator: '.image-tag', contains: ImageTag do
+        def view!
+          find('.js-image-tag-link').click
+          window.change_to(PageObjects::Images::IndexPage)
+        end
+      end
+
+      def tags_with_count
+        [tag_list.map(&:name), tag_list.map(&:count)].transpose
```
Comment (rmacklin) [inline test/page_objects/tags/index_page.rb]: This is a pretty neat implementation. It does `map` over the same collection twice and then perform a transpose, though, so there is definitely going to be redundant iteration which is something you should always look to avoid if it doesn't sacrifice readability. In this case, you can iterate over the collection just once without losing readability using:

```ruby
tag_list.map { |tag| [tag.name, tag.count] }
```

File: test/page_objects/images/new_page.rb [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+      form_for :image do
+        element :link
+        element :tag_list
+      end
+
+      def create_image!(link: nil, tags: nil)
+        self.link.set link if link.present?
+        tag_list.set tags if tags.present?
+        node.click_button('Create Image')
+        window.change_to(ShowPage, NewPage)
```
Comment (rmacklin) [inline test/page_objects/images/new_page.rb]: Typically this is done as
```ruby
window.change_to(ShowPage, self.class)
```
which indicates "this page is either going to change to the show page, or it'll rerender this page again". Additionally, if you were to rename `NewPage`, you wouldn't have to modify this line.

The same comment applies to the other page objects that have themselves as an argument to `window.change_to`

File: test/page_objects/tags/index_page.rb (hunk 22 above)
Comment (lsq921224) [inline test/page_objects/tags/index_page.rb]: This is great!

**Comment by lsq921224** [comment]
@rmacklin I have updated the PR.
